### PR TITLE
Feature/parser working

### PR DESCRIPTION
**Pull request by iRhysBradbury**
Welcome to the worlds largest PR.
~100k lines are actually:

- `Tokens.scala` `Seq[String]`
- Test resources which need to be shipped with the codebase 
- The `libbackend.so`
- Some tests are ignored because their expected have changed. Will fix later.

~5k lines are actual code to review.

- The `CrfScala` interface has already been reviewed.
- The implementation of the interface needs to be reviewed.
- There's a lot of string construction stuff going on, please refrain from commenting, some of it isnt nice, but I have optimisations in mind.

Optimisations (its pretty fast anyway):
- Impl a proper way of configuring AggregateFeatureAnalysers (current impl is because of time)
- Some places could do with some extra StringBuilders (specifically the AggregateFeatureAnalysers) 
- Static Types on Tokens.

Tests which need to be added:
- Specific token bounds test (40 odd to come)
- AggregateFeatureAnalysers when impl complete.

A special thanks to @Mironor @alangorton and @saniemi for your assistance with this.

### CODE REVIEW

**Comment by iRhysBradbury** [comment]
This branch will be squashed when merged.

File: parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfAggregateFeature.scala [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+  /**
+    * @return a function which returns an instance of T
+    */
+  def analyser(): CrfAggregateFeatureAnalyser[T]
+
+  /**
+    * @return name
+    */
+  def name(): String
```
Comment (alexflav23) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfAggregateFeature.scala]: `()` is by consensus an indicator of side effects, if there are none you should prefer the simple variant: `def name: String`.

File: parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfFeature.scala [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+  /**
+    * @return a function which returns an instance of T
+    */
+  def analyser() : CrfFeatureAnalyser[T]
+
+  /**
+    * @return name
+    */
+  def name() : String
```
Comment (alexflav23) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfFeature.scala]: Same as above, a bit of duplication in traits as well that could be delegated to a source `trait`, e.g something that both these classes extend.

File: parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfFeature.scala [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+/** todo scaladoc
+  *
+  * @tparam T the return type of the FeatureAnalyser
+  */
+trait CrfFeature[T] {
+
+  /**
+    * @return a function which returns an instance of T
+    */
+  def analyser() : CrfFeatureAnalyser[T]
```
Comment (alexflav23) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfFeature.scala]: Same feedback for `()`.

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/AddressParserMainTest.scala [hunk 4]
```diff
@@ -0,0 +1,132 @@
+package uk.gov.ons.addressIndex.parsers
+
+import java.io.File
+import uk.gov.ons.addressIndex.crfscala.CrfScalaJni
+import scala.util.Try
+import scala.xml.XML
+
+//Forgive me father for I have sinned.
+object AddressParserMainTest extends App {
```
Comment (Mironor) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/AddressParserMainTest.scala]: Does this code need to be reviewed or it's just a temporary usages?

File: parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfParser.scala [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        _.split(CrfScalaJni.delimiter)
+         .sorted
+         .mkString(CrfScalaJni.delimiter)
+      )
+      .mkString(CrfScalaJni.lineEnd) + CrfScalaJni.lineEnd
+    crfScala.tag(modelPath, augmentedActual)
+
+  }
+
+  //TODO scaladoc
```
Comment (alexflav23) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfParser.scala]: This is like a new level of procrastination

File: parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfParser.scala [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+  def parse(i: Input, fas: CrfFeatures, tokenable: CrfTokenable): CrfJniInput = {
+    val tokens = tokenable(i)
+    val preprocessedTokens = tokenable normalise tokens
+    val onlyOneToken = preprocessedTokens.length == 1
+    val onlyTwoTokens = preprocessedTokens.length == 2
+    val multipleTokens = preprocessedTokens.length > 2
+    val sb = StringBuilder.newBuilder
+
+    if(onlyOneToken) {
```
Comment (alexflav23) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfParser.scala]: Space after `if`, should be `if (`

File: parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfParser.scala [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        if(i == 0) {
+          sb
+            .append(
+              fas.toCrfJniInput(
+                input = preprocessedTokens(i),
+                next = Some(preprocessedTokens(i + 1))
+              ).replace("\n", "") //todo remove when aggr impl done
+              + "\trawstring.start:1.0\n"//todo impl AggregateFeatureAnalysers
+            )
+        } else if(i != preprocessedTokens.length - 1) {
```
Comment (alexflav23) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfParser.scala]: Space after `if` everywhere in this block

File: parsers/src/main/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysers.scala [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+        s"w:$length"
+      }
+    }
+  }
+
+  val endsInPunctuation: FeatureName = "endsinpunc"
+  /**
+    * @return true if the last character of the string is a '.', false if not
+    */
+  def endsInPunctuationAnalyser(): CrfFeatureAnalyser[Boolean] = CrfFeatureAnalyser[Boolean](_.last == '.')
```
Comment (alexflav23) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysers.scala]: A lot of these accessors are not modifiers and have no side effect so you can remove `()`.

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/AddressParserTest.scala [hunk 9]
```diff
@@ -1,328 +1,31 @@
...
-      "CYF",
-      "PLC",
-      "CCC",
-      "UNLTD",
-      "ULTD"
-    )
-    val expected = Seq.fill(inputs.length)(true)
-    val actual = inputs map companyAnalyser.apply
-    expected should contain theSameElementsAs actual
+  ignore should "create the correct `CrfJniInput` for input `mySingleton`" in {
```
Comment (alexflav23) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/AddressParserTest.scala]: Why are some tests still ignored?

File: server/conf/routes [hunk 10]
```diff
@@ -7,4 +7,4 @@ GET     /addresses          uk.gov.ons.addressIndex.server.controllers.AddressCo
 GET     /addresses/:uprn    uk.gov.ons.addressIndex.server.controllers.AddressController.uprnQuery(uprn, format)
 GET     /es                 uk.gov.ons.addressIndex.server.controllers.AddressController.elasticTest
 #swagger play
-#GET     /api                controllers.ApiHelpController.getResources
\ No newline at end of file
+#GET    /api                controllers.ApiHelpController.getResources
```
Comment (alexflav23) [inline server/conf/routes]: Still commented out?

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/AddressParserTest.scala (hunk 9 above)
Comment (Mironor) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/AddressParserTest.scala]: Either un-ignore it or remove it

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/AddressParserTest.scala [hunk 11]
```diff
@@ -1,328 +1,31 @@
...
-      "GARDENS",
-      "GROVE",
-      "VIEW",
-      "HILL",
-      "GREEN"
-    )
-    val expected = Seq.fill(inputs.length)(true)
-    val actual = inputs map roadAnalyser.apply
-    expected should contain theSameElementsAs actual
+  ignore should "create the correct `CrfJniInput` for input `my pair`" in {
```
Comment (Mironor) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/AddressParserTest.scala]: Either un-ignore it or remove it

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/AddressParserTest.scala [hunk 12]
```diff
@@ -1,328 +1,31 @@
...
-    val expected = Seq.fill(inputs.length)(true)
-    val actual = inputs map ordinalAnalyser.apply
-    expected should contain theSameElementsAs actual
-  }
-
-  it should "produce the correct `hyphenations` feature output for token `my-road`" in {
-    val input = "my-road"
-    val expected = 1
-    val actual = hyphenationsAnalyser apply input
+  ignore should "create the correct `CrfJniInput` for input `Surbiton`" in {
```
Comment (Mironor) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/AddressParserTest.scala]: Either un-ignore it or remove it

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala [hunk 13]
```diff
@@ -0,0 +1,334 @@
...
+    expected shouldBe actual
+  }
+
+  it should "have a `DigitsLiteral` for `noDigits` of `no_digits`" in {
+    val expected = "no_digits"
+    val actual = DigitsLiteral.noDigits
+    expected shouldBe actual
+  }
+
+  it should "have all feature names" in {
```
Comment (Mironor) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala]: This spec is testing the cast from String to Feature?

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala [hunk 14]
```diff
@@ -0,0 +1,334 @@
...
+      "EAST",
+      "WEST",
+      "NORTHEAST",
+      "NORTHWEST",
+      "SOUTHEAST",
+      "SOUTHWEST"
+    )
+    val expected = Seq.fill(inputs.length)(true)
+    val actual = inputs map directionalAnalyser.apply
+    expected should contain theSameElementsAs actual
```
Comment (Mironor) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala]: `shouldBe actual` will be fine in this case

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala [hunk 15]
```diff
@@ -0,0 +1,334 @@
...
+      "FLAT", "FLT",
+      "APARTMENT", "APPTS", "APPT", "APTS", "APT",
+      "ROOM",
+      "ANNEX",  "ANNEXE",
+      "UNIT",
+      "BLOCK", "BLK"
+    )
+    val expected = Seq.fill(inputs.length)(true)
+    val actual = inputs map flatAnalyser.apply
+    expected should contain theSameElementsAs actual
```
Comment (Mironor) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala]: `shouldBe actual` will be fine here

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala [hunk 16]
```diff
@@ -0,0 +1,334 @@
...
+      "LIMITED",
+      "CYF",
+      "PLC",
+      "CCC",
+      "UNLTD",
+      "ULTD"
+    )
+    val expected = Seq.fill(inputs.length)(true)
+    val actual = inputs map companyAnalyser.apply
+    expected should contain theSameElementsAs actual
```
Comment (Mironor) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala]: `shouldBe actual` will be fine here

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala [hunk 17]
```diff
@@ -0,0 +1,334 @@
...
+      "PLACE",
+      "GARDENS",
+      "GROVE",
+      "VIEW",
+      "HILL",
+      "GREEN"
+    )
+    val expected = Seq.fill(inputs.length)(true)
+    val actual = inputs map roadAnalyser.apply
+    expected should contain theSameElementsAs actual
```
Comment (Mironor) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala]: `shouldBe actual` will be fine here

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala [hunk 18]
```diff
@@ -0,0 +1,334 @@
...
+      "FARM",
+      "LODGE",
+      "COTTAGE", "COTTAGES",
+      "VILLA", "VILLAS",
+      "MAISONETTE",
+      "MEWS"
+    )
+    val expected = Seq.fill(inputs.length)(true)
+    val actual = inputs map residentialAnalyser.apply
+    expected should contain theSameElementsAs actual
```
Comment (Mironor) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala]: `shouldBe actual` will be fine here

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala [hunk 19]
```diff
@@ -0,0 +1,334 @@
...
+      "UWE",
+      "UEA",
+      "LSE",
+      "KCL",
+      "UCL",
+      "UNI", "UNIV", "UNIVERSITY", "UNIVERISTY"
+    )
+    val expected = Seq.fill(inputs.length)(true)
+    val actual = inputs map businessAnalyser.apply
+    expected should contain theSameElementsAs actual
```
Comment (Mironor) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala]: `shouldBe actual` will be fine here

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala [hunk 20]
```diff
@@ -0,0 +1,334 @@
...
+  it should "produce the correct `locational` feature output for expected inputs" in {
+    val inputs = Seq(
+      "BASEMENT", "GROUND", "ATTIC",
+      "UPPER", "ABOVE", "TOP", "LOWER", "FLOOR", "HIGHER",
+      "LEFT", "RIGHT", "FRONT", "BACK", "REAR",
+      "WHOLE", "PART", "SIDE"
+    )
+    val expected = Seq.fill(inputs.length)(true)
+    val actual = inputs map locationalAnalyser.apply
+    expected should contain theSameElementsAs actual
```
Comment (Mironor) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala]: `shouldBe actual` will be fine here

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala [hunk 21]
```diff
@@ -0,0 +1,334 @@
...
+      "THIRD", "3RD",
+      "FOURTH", "4TH",
+      "FIFTH", "5TH",
+      "SIXTH", "6TH",
+      "SEVENTH", "7TH",
+      "EIGHT", "8TH"
+    )
+    val expected = Seq.fill(inputs.length)(true)
+    val actual = inputs map ordinalAnalyser.apply
+    expected should contain theSameElementsAs actual
```
Comment (Mironor) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala]: `shouldBe actual` will be fine here

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/TokensTest.scala [hunk 22]
```diff
@@ -24,7 +32,7 @@ class TokensTest extends FlatSpec with Matchers {
       "TownName",
       "Postcode"
     )
-    val actual : Seq[Token]  = Tokens.ALL
+    val actual : Seq[Token]  = Tokens.all
```
Comment (Mironor) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/TokensTest.scala]: Why lowercase it?

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/TokensTest.scala [hunk 23]
```diff
@@ -172,19 +180,47 @@ class TokensTest extends FlatSpec with Matchers {
...
-    val actual : Seq[TokenIndicator] = Tokens.POST_TOWN
+    val actual : Seq[TokenIndicator] = Tokens.postTown
     expected should contain theSameElementsAs actual
   }
+
+  it should "convert a string to tokens separating on ` `" in {
+    val expected = List("one", "two", "three")
+    val input = expected.mkString(" ")
+    val actual = Tokens(input)
+    actual should contain theSameElementsAs expected
```
Comment (Mironor) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/TokensTest.scala]: Why not `shouldBe expected` ?

File: parsers/src/main/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysers.scala (hunk 8 above)
Comment (Mironor) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysers.scala]: I'd say that most of them can be `val`

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/AddressParserMainTest.scala (hunk 4 above)
Comment (Mironor) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/AddressParserMainTest.scala]: Because it needs a lot of refactoring, but maybe it will be removed shortly, so refactoring it will be pointless

File: parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfParser.scala (hunk 6 above)
Comment (iRhysBradbury) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfParser.scala]: This is a bit of a hack due to time. I have a Aggregate Feature Analyser branch which solves all this

File: parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfParser.scala [hunk 24]
```diff
@@ -0,0 +1,120 @@
...
+  }
+
+  //TODO scaladoc
+  def parse(i: Input, fas: CrfFeatures, tokenable: CrfTokenable): CrfJniInput = {
+    val tokens = tokenable(i)
+    val preprocessedTokens = tokenable normalise tokens
+    val onlyOneToken = preprocessedTokens.length == 1
+    val onlyTwoTokens = preprocessedTokens.length == 2
+    val multipleTokens = preprocessedTokens.length > 2
+    val sb = StringBuilder.newBuilder
```
Comment (iRhysBradbury) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfParser.scala]: StringBuilder is a must

File: parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfFeature.scala [hunk 25]
```diff
@@ -0,0 +1,105 @@
...
+    *
+    * @param input
+    * @param next
+    * @param previous
+    * @return
+    */
+  def toCrfJniInput(input: CrfToken, next: Option[CrfToken] = None, previous: Option[CrfToken] = None): CrfJniInput = {
+    StringBuilder
+      .newBuilder
+      .append(CrfScalaJni.lineStart)
```
Comment (iRhysBradbury) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfFeature.scala]: A string builder here is a must

File: parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfFeature.scala (hunk 2 above)
Comment (iRhysBradbury) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfFeature.scala]: I have a working branch which fixes this, to come

File: parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfAggregateFeatureAnalyser.scala [hunk 26]
```diff
@@ -0,0 +1,23 @@
...
+  /**
+    * Helper apply method for better syntax.
+    * Constructs a function.
+    * Eg Impl:
+    *
+    *    CrfAggregateFeatureAnalyser[Int]("indexOfToken") { (tokens, token) =>
+    *       tokens.indexOf(token)
+    *    }
+    */
+  def apply[T](analyser: CrfAggregateFeatureAnalyser[T]): CrfAggregateFeatureAnalyser[T] = analyser
```
Comment (iRhysBradbury) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfAggregateFeatureAnalyser.scala]: Ahh, the javadoc is a little off

File: parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfScala.scala [hunk 27]
```diff
@@ -1,147 +1,45 @@
...
-import com.github.jcrfsuite.CrfTagger
-import third_party.org.chokkan.crfsuite.{Attribute, Item}
-import uk.gov.ons.addressIndex.parsers.Tokens
-
-import scala.util.control.NonFatal
-
 /**
   * scala wrapper of crfsuite
+  *
+  * todo describe this more
```
Comment (iRhysBradbury) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfScala.scala]: +1

File: parsers/src/test/scala/uk/gov/ons/addressIndex/crfscala/CrfFeatureTest.scala [hunk 28]
```diff
@@ -0,0 +1,120 @@
...
+    type tType = ArbType
+    val output: tType = ArbType()
+    val analyser = CrfFeatureAnalyser[tType](_ => output)
+    val name = "name"
+    val feature = CrfFeatureTestImpl[tType](name)(analyser)
+  }
+
+  it should "throw an `UnsupportedOperationException` for any type other than `Boolean`, `String`, `Int`, `Double` for input `arbitrary`" in {
+    val input = "arbitrary"
+    Try[String](TestInstanceArbType.feature.toCrfJniInput(input)) recover {
```
Comment (iRhysBradbury) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/crfscala/CrfFeatureTest.scala]: OoOo Thanks!

**Comment by iRhysBradbury** [comment]
Lots to do!

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala (hunk 14 above)
Comment (iRhysBradbury) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala]: Im testing for element equality not ordering

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala (hunk 15 above)
Comment (iRhysBradbury) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala]: Im testing for element equality not ordering

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala (hunk 16 above)
Comment (iRhysBradbury) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala]: Im testing for element equality not ordering

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala (hunk 17 above)
Comment (iRhysBradbury) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala]: Im testing for element equality not ordering

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala (hunk 18 above)
Comment (iRhysBradbury) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala]: Im testing for element equality not ordering

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala (hunk 21 above)
Comment (iRhysBradbury) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala]: Im testing for element equality not ordering

**Comment by iRhysBradbury** [comment]
Some tes feedback

File: parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfAggregateFeatureAnalyser.scala (hunk 26 above)
Comment (iRhysBradbury) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfAggregateFeatureAnalyser.scala]: This is for DSL purposes

File: parsers/src/test/scala/uk/gov/ons/addressIndex/crfscala/CrfFeatureTest.scala [hunk 29]
```diff
@@ -0,0 +1,120 @@
+package uk.gov.ons.addressIndex.crfscala
+
+import org.scalatest.{FlatSpec, Matchers}
+import uk.gov.ons.addressIndex.crfscala.CrfFeatureAnalyser.CrfFeatureAnalyser
+import scala.util.Try
+
+class CrfFeatureTest extends FlatSpec with Matchers  {
+
+  object TestInstanceBoolean {
+    type tType = Boolean
```
Comment (iRhysBradbury) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/crfscala/CrfFeatureTest.scala]: This is a little helper for the test; the idea is that i could C+P the block for the next type test and just change the type and value

File: parsers/src/test/scala/uk/gov/ons/addressIndex/crfscala/CrfFeatureTest.scala [hunk 30]
```diff
@@ -0,0 +1,120 @@
...
+    val analyser = CrfFeatureAnalyser[tType](_ => true)
+    val name = "name"
+    val feature = CrfFeatureTestImpl[tType](name)(analyser)
+  }
+
+  it should "return the same type as the type which is returned from its analyser for input `arbitrary`" in {
+    val input = "arbitrary"
+    val actual = TestInstanceBoolean.feature.analyse(input)
+    assert(actual.isInstanceOf[TestInstanceBoolean.tType])
```
Comment (iRhysBradbury) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/crfscala/CrfFeatureTest.scala]: Thanks

File: parsers/src/test/scala/uk/gov/ons/addressIndex/crfscala/CrfFeatureTest.scala [hunk 31]
```diff
@@ -0,0 +1,120 @@
...
+  }
+
+  it should "return the expected `CrfJniInput` for type `String` for input `arbitrary`" in {
+    val input = "arbitrary"
+    val actual = TestInstanceStringA.feature.toCrfJniInput(input)
+    val expected = s"\t${TestInstanceStringA.name}\\:$input:1.0"
+    expected shouldBe actual
+  }
+
+  object TestInstanceStringB {
```
Comment (iRhysBradbury) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/crfscala/CrfFeatureTest.scala]: Is this PR blocking?

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/AddressParserTest.scala (hunk 12 above)
Comment (iRhysBradbury) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/AddressParserTest.scala]: added scaladoc

**Comment by iRhysBradbury** [comment]
Adding comments on feedback

File: parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfFeatureAnalyser.scala [hunk 32]
```diff
@@ -0,0 +1,25 @@
...
+    *    }
+    *
+    * Or:
+    *
+    *    CrfFeatureAnalyser[Int]("lengthOfString") { str =>
+    *       str.length
+    *    }
+    *
+    */
+  def apply[T](analyser : CrfFeatureAnalyser[T]) : CrfFeatureAnalyser[T] = analyser
```
Comment (iRhysBradbury) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfFeatureAnalyser.scala]: This is for DSL impl

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/AddressParserMainTest.scala (hunk 4 above)
Comment (iRhysBradbury) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/AddressParserMainTest.scala]: Could be removed I suppose

File: parsers/src/main/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysers.scala [hunk 33]
```diff
@@ -0,0 +1,224 @@
...
+        }
+      }
+    }
+  }
+
+  /**
+    * ref digitsAnalyser
+    */
+  object DigitsLiteral {
+    val allDigits = "all_digits"
```
Comment (Mironor) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysers.scala]: For future reference it's neither ALL CAPS, nor camelCase: http://stackoverflow.com/a/9752606/555355

File: parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfFeature.scala [hunk 34]
```diff
@@ -34,34 +34,29 @@ trait CrfFeature[T] {
...
-      )
-      .append(
-        next map { next =>
-          CrfScalaJni.delimiter +
+    val currentCrfJni: CrfJniInput = createCrfJniInput(
+      prefix = name,
+      value = analyse(input)
+    )
+    val nextCrfJni: CrfJniInput = {
+      next map { next =>
```
Comment (alexflav23) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfFeature.scala]: Naming could improve if you are duplicating `next` it's hard to tell instantly which is the monad and which is the value.

File: parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfFeature.scala [hunk 35]
```diff
@@ -34,34 +34,29 @@ trait CrfFeature[T] {
     * @return
     */
   def toCrfJniInput(input: CrfToken, next: Option[CrfToken] = None, previous: Option[CrfToken] = None): CrfJniInput = {
```
Comment (alexflav23) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfFeature.scala]: I'd expect some really good Scala docs for a method this esoteric. There's nothing complex going on per say, but it's pretty hard to follow, speaking for myself as an external reader. I know it's already in the bucket list, and today is just a marvellous opportunity to outshine Dickens and earn the first Nobel prize for literature for a Scaladoc comment.

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala (hunk 19 above)
Comment (Mironor) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysersTest.scala]: For future reference, you're comparing expected with a List filled with `true` so ordering is not in question

File: parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfFeature.scala [hunk 36]
```diff
@@ -34,34 +34,29 @@ trait CrfFeature[T] {
...
           )
-        } getOrElse ""
-      )
-      .append(
-        previous map { previous =>
-          CrfScalaJni.delimiter +
+      } getOrElse ""
+    }
+    val previousCrfJni: CrfJniInput = {
+      previous map { previous =>
```
Comment (alexflav23) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfFeature.scala]: Same comment as above, make it obvious which is the monad and which is the element.

File: parsers/src/test/scala/uk/gov/ons/addressIndex/crfscala/CrfFeatureTest.scala (hunk 31 above)
Comment (Mironor) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/crfscala/CrfFeatureTest.scala]: No, just want to know, maybe I'm missing something

File: parsers/src/test/scala/uk/gov/ons/addressIndex/crfscala/CrfFeatureTest.scala (hunk 29 above)
Comment (alexflav23) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/crfscala/CrfFeatureTest.scala]: @iRhysBradbury I agree with @Mironor that this should be type classes, not duplicated. I'd expect to see a `trait TestInstance[TypeParam]` which you could simply extend, maybe even with implicit type class instances that you could materialize as context bounds.


```scala
trait TestInstance[T] {
 // ....
 def pullKateUpton(name: String, ambition: NumericRange[Long]): Try[T] = ...
}

object TestInstance {
   // this is just a convenience method.
   def apply[T : TestInstance]: TestInstance[T] = implicitly[TestInstance[T]]

   // it's equivalent to this basically
   def apply[T]()(implicit ev: TestInstance[T]): TestInstance[T] = ev

   implicit object TestInstanceBool extends TestInstance[Boolean] {
     override def pullKateUpton(name: String, ambition: NumericRange[Long]: Try[Boolean] = {
       if ((name == "Flavian" || name == "Rhys") && ambition.contains(10L))  {
          Success(true)
       } else {
          throw new Error("How dare you even try"))
       }
     }
   }
}
```
// now this is pretty fun since it works like magic.

```scala
val testInstance = TestInstance[Boolean]

testInstance.pullKateUpton("rhys", 1 to 5L) // false
testInstance.pullKateUpton("rhys", 1 to 10L) // TRUE !!! send pics please
```

File: parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfScalaJni.scala [hunk 37]
```diff
@@ -0,0 +1,32 @@
...
+    *
+    * @param modelPath
+    * @param items
+    * @return
+    */
+  @native def tag(modelPath: String, items : String) : String
+}
+
+object CrfScalaJni {
+  val tab = "\t"
```
Comment (alexflav23) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/crfscala/CrfScalaJni.scala]: @Mironor This is not a Scala default, constants are merely capitalised, but not entirely uppercased in scala. `Constant` not `CONSTANT`, by convention.
